### PR TITLE
api: json api reorganization and app_config

### PR DESCRIPTION
**Pull request by montagao**
This patch moves all of the JSON related functionality definitions to iotutils (under app_json), in order to reduce dependencies on the iotlib for the auxiliary apps that use the iotlib only for JSON parsing. iot_base (in the iotlib) still uses a wrapper around app_json called iot_json.

It also reintroduces app_config which is another wrapper around app_json to help parse config files specifically and to allow for a simpler interface for JSON parsing in the future, since several of the library tools require config reading functionality.

### CODE REVIEW

File: src/utilities/app_config.c [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+ *
+ * @param[out]     file_path           path to connection file
+ * @param[in]      len                 length of buffer
+ *
+ * @retval IOT_STATUS_BAD_PARAMETER    bad parameters
+ * @retval IOT_STATUS_FAILURE          on failure
+ * @retval IOT_STATUS_SUCCESS          on success
+ */
+static iot_status_t app_config_get_proxy_file_path(
+char *file_path, size_t len );
```
Comment (lt-holman) [inline src/utilities/app_config.c]: add tab to start of line.

File: src/utilities/app_config.c [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+
+iot_status_t app_config_close( struct app_config *config )
+{
+	iot_status_t result = IOT_STATUS_BAD_PARAMETER;
+	if ( config )
+	{
+		app_json_decode_terminate( config->json );
+		os_free( (void *)config );
```
Comment (lt-holman) [inline src/utilities/app_config.c]: shouldn't need to cast to "void *"

File: src/utilities/app_config.c [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+		result = IOT_STATUS_SUCCESS;
+	}
+	return result;
+}
+
+struct app_config *app_config_open(/* iot_t *iot_lib,*/ const char *file_path )
+{
+	// TODO: We only use iot_lib for logging purposes.
+	// should we remove it to reduce dependencies?
```
Comment (lt-holman) [inline src/utilities/app_config.c]: I agree with your comment here.

File: src/utilities/app_json.h [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied."
+ */
+#ifndef APP_JSON_H
+#define APP_JSON_H
+
+#include "api/public/iot.h"
```
Comment (lt-holman) [inline src/utilities/app_json.h]: why is this needed?

File: src/utilities/app_json.h [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+
+/* BASE SUPPORT */
+#ifndef IOT_STACK_ONLY
+/**
+ * @brief retrieves the function for dynamic memory allocation
+ *
+ * @param[out]     mptr                pointer to the realloction function
+ * @param[out]     fptr                pointer to the free function
+ */
+IOT_API IOT_SECTION void app_json_allocation_get(
```
Comment (lt-holman) [inline src/utilities/app_json.h]: Remove IOT_API... not part of the API (for all functions declared here).

File: src/utilities/app_json.h (hunk 4 above)
Comment (lt-holman) [reply src/utilities/app_json.h]: is it for IOT_STACK_ONLY... include "iot_build.h" then.  The point of this change is to use json functions with the including the API library.

File: src/utilities/app_json_base.h [hunk 6]
```diff
@@ -0,0 +1,91 @@
+/**
+ * @file
+ * @brief header file for base IoT library json functionality
+ *
+ * @copyright Copyright (C) 2017 Wind River Systems, Inc. All Rights Reserved.
```
Comment (lt-holman) [inline src/utilities/app_json_base.h]: all changed files should have a copyright of "2018"  (or add "-2018", if copyrighted from a previous year)

File: src/relay/relay_main.c [hunk 7]
```diff
@@ -44,10 +44,11 @@
 
 #include "iot_build.h"                 /* for IOT_NAME_FULL */
 #include "iot.h"                       /* for iot_bool_t type */
+#include "iot_mqtt.h"
 #include "shared/iot_types.h"          /* for proxy struct */
-#include "app_arg.h"                   /* for struct app_arg & functions */
-/*#include "app-common/app_config.h"     *//* for proxy configuration */
-/*#include "app-common/app_path.h"       *//* for path support functions */
+#include "../utilities/app_arg.h"      /* for struct app_arg & functions */
```
Comment (lt-holman) [inline src/relay/relay_main.c]: no ...

File: src/utilities/app_config.c [hunk 8]
```diff
@@ -14,8 +14,8 @@
 #include "app_config.h"
 
 #include <os.h>				/* for os_* functions */
-#include "app_path.h"		/* for app_path_* functions */
 #include "app_json.h"		/* for app_json_* functions */
+#include "app_path.h"		/* for app_path_* functions */
 #include "../public/iot_json.h"		/* for iot_json_* types */
```
Comment (lt-holman) [inline src/utilities/app_config.c]: app config... shouldn't need iot_json.h anymore.. app_json.h (unless this is ordered after the 1st commit)

**Comment by montagao** [comment]
merged the commits and addressed the requested changes. 
Currently looking at testing android and vxworks* builds.

*(might need some guidance for vxworks)
